### ISSUE TITLE
CRI: Should we checkpoint SandboxConfig and ContainerConfig and in dockershim?

### ISSUE THREAD

**Issue by Random-Liu** [kubernetes/kubernetes#34672]
The philosophy of current CRI is to only pass configuration (SandboxConfig, ContainerConfig and ImageSpec) when creating sandbox, container and image, and only use the reference (id) to manage them after creation.

However, in dockershim, the config is not only needed in the creation function:
- **We may need the config in other CRI functions.** E.g. the container log is created by docker only after the container is started. So we can only symlink the log files in `StartContainer` instead of `CreateContainer`. However, we have no information except container id in `StartContainer`.
- **We may need the config in a periodically sync.** Even though the CRI is imperative, we may still need a sync loop inside shim to guarantee something which docker can't do for us, such as network related stuff. The config should be check-pointed somewhere, so that the shim can recover state after restart.

If we checkpoint things, shoud we checkpoint the whole config or just fields we need?
- **The whole config.**
  - Pros: We can use the versioning of the api object directly. (We don't have object versioning yet)
  - Cons: There may be many thing we'll never use.
- **Only necessary fields.**
  - Pros: Clean.
  - Cons: We need to manage the checkpoint object separately.

If we checkpoint the whole config, should we checkpoint in a file or docker label?
- **Write the configuration into files.** Extra complexity. The checkpoint lifecycle is independent from container lifecycle, the shim has to manage the lifecycle itself.
- **Write the configuration into labels.** The checkpoint lifecycle is bound with container lifecycle.

@yujuhong @feiskyer 
/cc @kubernetes/sig-node

**Comment by feiskyer** [comment]
+1 for checkpointing config by labels for various use cases. 

> We may need the config in a periodically sync. Even though the CRI is imperative, we may still need a sync loop inside shim to guarantee something which docker can't do for us, such as network related stuff. The config should be check-pointed somewhere, so that the shim can recover state after restart.

Not sure why a periodically sync is needed. @Random-Liu Would you mind explain this more clearly?

**Comment by Random-Liu** [comment]
> Not sure why a periodically sync is needed. @Random-Liu Would you mind explain this more clearly?

In fact, this is not clear yet. Because dockershim is managing some resources which is not directly managed by docker, such as pod IP, log files and maybe more in the future. For example, if a sandbox is removed, the shim must guarantee the corresponding IP should be released. However, even though we put the ip release logic into `RemoveSandbox`, it may fail. Once it fails, we need a periodically sync to make sure the orphaned resources are released correctly.

There are 2 solutions for this:
1) Extend the container and sandbox state to make it be able to reflect whether all resources are released properly. If not, Kubelet will retry. (@yujuhong is working on this now)
2) Delegate these logic to shim, and assume that shim can always make sure resources are released properly once container/sandbox is stopped/removed.

**Comment by yifan-gu** [comment]
@Random-Liu Can we clean up all resources before actually removing the container? And if there are any errors, try later.

**Comment by yifan-gu** [comment]
> We may need the config in other CRI functions

You can get the information from labels, right?

**Comment by euank** [comment]
>  1) Extend the container and sandbox state to make it be able to reflect whether all resources are released properly. If not, Kubelet will retry. (@yujuhong is thinking about this now)

I don't see why we'd need more status codes in that case. The only thing kubelet needs to care about is if it succeeded or not, the runtime then just needs to be able to figure out what it's already done itself, but that all seems like it can be internal

**Comment by Random-Liu** [comment]
> Can we clean up all resources before actually removing the container? And if there are any errors, try later.

> I don't see why we'd need more status codes in that case. The only thing kubelet needs to care about is if it succeeded or not, the runtime then just needs to be able to figure out what it's already done itself, but that all seems like it can be internal

One example is `StopContainer`. Different runtime may have different implementation for `StopContainer`. If `StopContainer` does release some resource, e.g. IP address, then once it fails, should the container be marked as `Running` or `Exited`?

If it is `Running`, I'm not sure whether Kubelet may get confused. We are not sure whether `Running` means a working container or just some leaked resource.
If it is `Exited`, then Kubelet will never try to clean up it again.

**Comment by yifan-gu** [comment]
> If StopContainer does release some resource, e.g. IP address, then once it fails, should the container be marked as Running or Exited?

First, I am not sure which runtime is cleaning up IPs in `StopContainer`. Can we suggest runtime clean up resources in `RemoveContainer`? Why this is not achievable? It's true that different runtime will have different implementation, but isn't the whole purpose of the CRI is to make some explicit contract for runtimes? And I think we can make it clear by stating something like:

```
When a container is "stopped", no processes in the containers is running, but it might still hold some resources (IPs, mounts, etc) until "RemoveContainer" is called and succeeded.
```

**Comment by Random-Liu** [comment]
> First, I am not sure which runtime is cleaning up IPs in StopContainer? Can we suggest runtime clean up resources in RemoveContainer? Why this is not achievable?

@yifan-gu 
1) In current kubelet implementation, we'll not remove all containers unless the pod is deleted, so that user can still describe the pod and get container status. If these stopped but not removed container still hold scarce resources, like IP address, we'll have less and less available addresses unless the pods are deleted.
2) In the dockershim implementation, @freehan told me that we'll release IP address when stopping container.

**Comment by Random-Liu** [comment]
> I don't see why we'd need more status codes in that case. The only thing kubelet needs to care about is if it succeeded or not, the runtime then just needs to be able to figure out what it's already done itself, but that all seems like it can be internal

@euank Yeah, I agree. I believe rkt itself will manage all resources well, and kubelet don't need to worry about any resource leakage or something, just call CRI is enough.

For dockershim, because we are managing something which is not managed by docker directly, such as container logs, ip addresses and maybe more in the future. Ideally, dockershim itself should manage the resources well, and make `dockershim+docker` behave like `rkt` (`solution 2` in https://github.com/kubernetes/kubernetes/issues/34672#issuecomment-253870361).

However, another possible solution is explicitly add the `bad state` and let Kubelet handle this. (`solution 1` in https://github.com/kubernetes/kubernetes/issues/34672#issuecomment-253870361)

**Comment by yifan-gu** [comment]
@Random-Liu OK you are right, we teardown() before killsandbox(). (shame on me, actually I am the one who just wrote the exactly same code in the refactor PR...)

How about we also run teardown()/cleanup() before RemoveSandbox()?
Which is a no-op if it's already cleaned up, and if it fails, it blocks the RemoveSandbox().

**Comment by feiskyer** [comment]
> Because dockershim is managing some resources which is not directly managed by docker, such as pod IP, log files and maybe more in the future. For example, if a sandbox is removed, dockershim must guarantee the corresponding IP should be released. However, even though we put the ip release logic into RemoveSandbox, it may fail. Once it fails, we need a periodically sync to make sure the orphaned resources are released peroperly.

@Random-Liu  I think releasing resources is part of `RemoveSandbox`, e.g. if resources release failed, the api should report an error. Once `RemoveSandbox` failed, kubelet will continue to see this sandbox. And kubelet will call `RemoveSandbox` again later so shim could release the resources again. That's why I think periodically sync is not required.

> You can get the information from labels, right?

@yifan-gu We don't expect runtimes to understand labels written by kubelet. So if a runtime needs those configs somewhere, it needs checkpointing.

I think other runtimes also need to checkpoint those configs for releasing resources. This reminds me another thing ( @yifan-gu have same question before): since configs have already been checkpointed, do we need `SandboxConfig` in `CreateContainerRequest`?  cc/ @yujuhong

**Comment by yifan-gu** [comment]
> @Random-Liu I think releasing resources is part of RemoveSandbox, e.g. if resources release failed, the api should report an error. Once RemoveSandbox failed, kubelet will continue to see this sandbox. And kubelet will call RemoveSandbox again later so shim could release the resources again. That's why I think periodically sync is not required.

+1

That's why I am saying checkpointing config isn't necessary the solution for resource clean up.

**Comment by freehan** [comment]
Anyone start working on this?

I am working on enabling the hostport for CRI. This is one of the missing pieces. 

RunPodSandbox takes the PodSandboxConfig which includes containerPorts, but StopPodSandbox did not. That means either the network plugin or docker shim has to store it.

I prefer to checkpoint it in shim because this information is required across network plugins. 

I prefer not to store the containerPorts in docker labels because: 1. if the sandbox container went missing (user stop and delete it), there is no way we can retrieve this information without reaching out to kubelet. Then network plugin may leak iptables rules. 2. it is kind of clumsy to store a list of container ports config in one label, or we have to do multiple labels with variable keys. 


Any thoughts?

**Comment by freehan** [comment]
Ideally, I want to checkpoints stuff like: `netns`, `containerport`, `podcidr`
The structure may look like this:

```
/var/lib/dockershim/config/podCIDR
/var/lib/dockershim/sandbox/${SandboxID}/netns
/var/lib/dockershim/sandbox/${SandboxID}/ports/${port name or index}/hostport
/var/lib/dockershim/sandbox/${SandboxID}/ports/${port name or index}/containerport
/var/lib/dockershim/sandbox/${SandboxID}/ports/${port name or index}/protocol
/var/lib/dockershim/sandbox/${SandboxID}/ports/${port name or index}/hostIP
```

**Comment by thockin** [comment]
Just a couple quick thoughts:

* Do we want this under a whole new directory (/var/lib/dockershim) or /var/lib/kubelet/ somewhere?  I just want to make sure this is an intentional choice.

* Kubelet has, thus far, not really checkpointed things except as part of things it already had to write to disk (e.g. the path to a volume mount).  Think about how little we can get away with here.

* What are the values of these files?  What do they *mean* and what guarantees do they provide?  What happens if they don't exist or are corrupted?  How will you know if they are corrupted?

* Consider carefully the model of many single-value files vs a single multi-value file.  File writes have to be sync'ed to disk before you can really trust them, so we should capture that in a single lib.  Syncs are expensive.  How often do we expect any of these to be updated?  O(1) is OK.  Every 5 seconds maybe not.

* Will we check and re-write these in case they are removed or corrupted?

* Think about schema and versioning.  You are defining a serialization format - you'd better think hard about forward and backward compat, and updates and rollbacks of kubelet/dockershim.

Checkpointing is easy to get wrong.  Please please think of the many ways this could break.  What's the worst possible intermediate state where if kubelet crashed, we're screwed.  Remember the mantra: write first, act second.  Make sure the checkpoint is safe before you activate it.

**Comment by feiskyer** [comment]
@thockin Great summary of potential checkpointing problems.

> The structure may look like this:

@freehan Instead of writing them into multiple files, how about joining them in one file by json-encoding? So we only need one file sync for each sandbox.

**Comment by yujuhong** [comment]
We (@thockin, @freehan, @dchen1107)  had a discussion recently, since the network-related information is the main driver for checkpointing for now.

@dchen1107 brought up a good point that it might be too premature to commit to a generic checkpointing (either in kubelet or in the shim). Maintaining backward-compatibility would be a pain and could slow down the development in the future. Instead, we can work bottom-up, i.e., let components checkpoint the minimal, necessary information, to understand the needs, and eventually derive the scope and requirements for a more generic mechanism.

@freehan will work on checkpointing only the required information (e.g., sandbox ids and host ports) for networking in the dockershim.

**Comment by freehan** [comment]
This has been implemented.

**Comment by zhangguanzhang** [comment]
`pkg/kubelet/config/flags.go`
```
	fs.StringVar(&s.DockershimRootDirectory, "experimental-dockershim-root-directory", s.DockershimRootDirectory, "Path to the dockershim root directory.")
	fs.MarkHidden("experimental-dockershim-root-directory")
```
could use the option `--experimental-dockershim-root-directory` which had been hidden to change the `DockershimRootDirectory` from `/var/lib/dockershim` to others
